README.md
# playoff_predictor
Machine learning model to predict winners of first round of the 2022 nba playoffs.

**Project Summary: Predicting NBA Playoff Winners Using Machine Learning**

**Objective:**  
Predict the outcome of NBA playoff games using team performance metrics from the regular season (2022-2023).


## File explanation

os_helper.py - untitly module to move files around, see contents or zipped files, unzip bulk files.

API - API and directions to updating data or downloading data from the three sources. <https://github.com/shufinskiy/nba_apiv3/tree/master>

csv - file that contains all the relevant dataset


## Data Source
This dataset comprises NBA play-by-play data and shot details spanning from the 1996/1997 to 2022/23 seasons, with the latest update on 2023-07-11 incorporating playoff data for all seasons. The data is sourced from three platforms: 
* stats.nba.com 
* data.nba.com
* pbpstats.com 

Each of the datasets have a differnt start season and minor inconsistencies toughout the datasets. For this project we will only be using 2021-2022 , and 2022,2023 seanons and the first round of the playoff.

Stats not available or missing values were updated by api.



### nbastats


| Field                      | Description                                          |
|----------------------------|------------------------------------------------------|
| **GAME_ID**                | Game ID                                              |
| **EVENTNUM**               | Sequence number of the event                         |
| **EVENTMSGTYPE**           | Type of event (e.g., shot, rebound, foul)            |
| **EVENTMSGACTIONTYPE**     | Subtype or specific action of the event              |
| **PERIOD**                 | Quarter number                                       |
| **WCTIMESTRING**           | Wall-clock time when the event occurred              |
| **PCTIMESTRING**           | Time remaining in the quarter                        |
| **HOMEDESCRIPTION**        | Description of the home team action                  |
| **NEUTRALDESCRIPTION**     | Description of a neutral action (e.g., start of period) |
| **VISITORDESCRIPTION**     | Description of the visitor team action               |
| **SCORE**                  | Current game score at the time of the event          |
| **SCOREMARGIN**            | Score margin at the time of the event                |
| **PERSON1TYPE**            | Type identifier for the first person involved        |
| **PLAYER1_ID**             | ID of the player who performed the main action       |
| **PLAYER1_NAME**           | Name of the player who performed the main action     |
| **PLAYER1_TEAM_ID**        | Team ID of the player who performed the main action  |
| **PLAYER1_TEAM_CITY**      | Team city of the player who performed the main action|
| **PLAYER1_TEAM_NICKNAME**  | Team nickname of the player who performed the main action |
| **PLAYER1_TEAM_ABBREVIATION**| Team abbreviation of the player who performed the main action|
| **PERSON2TYPE**            | Type identifier for the second person involved       |
| **PLAYER2_ID**             | ID of the player who performed a side action         |
| **PLAYER2_NAME**           | Name of the player who performed a side action       |
| **PLAYER2_TEAM_ID**        | Team ID of the player who performed a side action    |
| **PLAYER2_TEAM_CITY**      | Team city of the player who performed a side action  |
| **PLAYER2_TEAM_NICKNAME**  | Team nickname of the player who performed a side action |
| **PLAYER2_TEAM_ABBREVIATION**| Team abbreviation of the player who performed a side action|
| **PERSON3TYPE**            | Type identifier for the third person involved        |
| **PLAYER3_ID**             | ID of the player who performed a second side action  |
| **PLAYER3_NAME**           | Name of the player who performed a second side action|
| **PLAYER3_TEAM_ID**        | Team ID of the player who performed a second side action|
| **PLAYER3_TEAM_CITY**      | Team city of the player who performed a second side action|
| **PLAYER3_TEAM_NICKNAME**  | Team nickname of the player who performed a second side action|
| **PLAYER3_TEAM_ABBREVIATION**| Team abbreviation of the player who performed a second side action|
| **VIDEO_AVAILABLE_FLAG**   | Indicates if a video of the event is available       |



### nbadata

| Field           | Description                                             |
|-----------------|---------------------------------------------------------|
| **evt**         | Sequence number of the event in the game                |
| **wallclk**     | Wall clock time when the event occurred                 |
| **cl**          | Time until end of the quarter                           |
| **de**          | Description of the action                               |
| **locX**        | Coordinate of shot along the width of the court relative to its central axis |
| **locY**        | Coordinate of shot along the length of the court relative to its central axis |
| **opt1**        | Points                                                  |
| **opt2**        | Additional option (context-dependent)                   |
| **opt3**        | Additional option (context-dependent)                   |
| **opt4**        | Additional option (context-dependent)                   |
| **tid**         | Team ID                                                 |
| **pid**         | Player ID                                               |
| **hs**          | Home team score                                         |
| **vs**          | Visitor team score                                      |
| **epid**        | Extra player ID (context-dependent)                     |
| **oftid**       | Team ID in offense                                      |
| **ord**         | Order number (context-dependent)                        |
| **pts**         | Points scored in the event                               |
| **PERIOD**      | Quarter number                                          |
| **GAME_ID**     | Game ID                                                 |


### pbpstats
| Field                           | Description                                              |
|---------------------------------|----------------------------------------------------------|
| **ENDTIME**                     | Time until end of the quarter at time of end of possession |
| **EVENTS**                      | Description of all actions in possession                 |
| **FG2A**                        | Count of 2PT Field Goal attempts in possession           |
| **FG2M**                        | Count of 2PT Field Goals made in possession              |
| **FG3A**                        | Count of 3PT Field Goal attempts in possession           |
| **FG3M**                        | Count of 3PT Field Goals made in possession              |
| **GAMEDATE**                    | Game date                                                |
| **GAMEID**                      | Game ID                                                  |
| **NONSHOOTINGFOULSTHATRESULTEDINFTS** | Non-shooting fouls that resulted in free throws    |
| **OFFENSIVEREBOUNDS**           | Count of offensive rebounds in possession                |
| **OPPONENT**                    | Abbreviation of the team in defense                      |
| **PERIOD**                      | Quarter number                                           |
| **SHOOTINGFOULSDRAWN**          | Shooting fouls drawn                                     |
| **STARTSCOREDIFFERENTIAL**      | Difference in score at the start of possession           |
| **STARTTIME**                   | Time until end of the quarter at time of start of possession |
| **STARTTYPE**                   | Type of start (context-dependent)                        |
| **TURNOVERS**                   | Turnovers                                                |
| **DESCRIPTION**                 | Description of action                                    |
| **URL**                         | Link to video                                            |



### datanba

| Field           | Description                                             |
|-----------------|---------------------------------------------------------|
| **evt**         | Sequence number of the event in the game                |
| **wallclk**     | Wall clock time when the event occurred                 |
| **cl**          | Time until end of the quarter                           |
| **de**          | Description of the action                               |
| **locX**        | Coordinate of shot along the width of the court relative to its central axis |
| **locY**        | Coordinate of shot along the length of the court relative to its central axis |
| **opt1**        | Points                                                  |
| **opt2**        | Additional option (context-dependent)                   |
| **mtype**       | Type of the move (context-dependent)                    |
| **etype**       | Type of the event (context-dependent)                   |
| **opid**        | Opponent player ID (context-dependent)                  |
| **tid**         | Team ID                                                 |
| **pid**         | Player ID                                               |
| **hs**          | Home team score                                         |
| **vs**          | Visitor team score                                      |
| **epid**        | Extra player ID (context-dependent)                     |
| **oftid**       | Team ID in offense                                      |
| **ord**         | Order number (context-dependent)                        |
| **PERIOD**      | Quarter number                                          |
| **GAME_ID**     | Game ID                                                 |




### shotdetail

| Field                 | Description                                               |
|-----------------------|-----------------------------------------------------------|
| **GRID_TYPE**         | Shot Chart Detail                                         |
| **GAME_ID**           | Game ID                                                   |
| **GAME_EVENT_ID**     | Sequence number of the event in the game                  |
| **PLAYER_ID**         | ID of player who made the shot                            |
| **PLAYER_NAME**       | Name of player who made the shot                          |
| **TEAM_ID**           | Team ID of player who made the shot                       |
| **TEAM_NAME**         | Team name of player who made the shot                     |
| **PERIOD**            | Quarter number                                            |
| **MINUTES_REMAINING** | Minutes remaining until end of the quarter                |
| **SECONDS_REMAINING** | Seconds remaining until end of the quarter                |
| **EVENT_TYPE**        | Made or Missed shot                                       |
| **ACTION_TYPE**       | Type of shot                                              |
| **SHOT_TYPE**         | 2PT or 3PT shot                                           |
| **SHOT_ZONE_BASIC**   | General area on the court where the shot was taken        |
| **SHOT_ZONE_AREA**    | Specific area on the court where the shot was taken       |
| **SHOT_ZONE_RANGE**   | Distance range of the shot                                |
| **SHOT_DISTANCE**     | Distance to the rim                                       |
| **LOC_X**             | Coordinate of the shot along the width of the court relative to its central axis |
| **LOC_Y**             | Coordinate of the shot along the length of the court relative to its central axis |
| **SHOT_ATTEMPTED_FLAG**| Shot execution flag (always 1)                           |
| **SHOT_MADE_FLAG**    | Shot made flag (0 or 1)                                   |
| **GAME_DATE**         | Game date                                                 |
| **HTM**               | Abbreviation of home team                                 |
| **VTM**               | Abbreviation of away team                                 |



| Team Name                 | team_id       |
|---------------------------|---------------|
| Atlanta Hawks            | **1610612737** |
| Boston Celtics           | **1610612738** |
| Brooklyn Nets            | **1610612751** |
| Charlotte Hornets        | **1610612766** |
| Chicago Bulls            | **1610612741** |
| Cleveland Cavaliers      | **1610612739** |
| Dallas Mavericks         | **1610612742** |
| Denver Nuggets           | **1610612743** |
| Detroit Pistons          | **1610612765** |
| Golden State Warriors    | **1610612744** |
| Houston Rockets          | **1610612745** |
| Indiana Pacers           | **1610612754** |
| LA Clippers              | **1610612746** |
| Los Angeles Lakers       | **1610612747** |
| Memphis Grizzlies        | **1610612763** |
| Miami Heat               | **1610612748** |
| Milwaukee Bucks          | **1610612749** |
| Minnesota Timberwolves   | **1610612750** |
| New Orleans Pelicans     | **1610612740** |
| New York Knicks          | **1610612752** |
| Oklahoma City Thunder    | **1610612760** |
| Orlando Magic            | **1610612753** |
| Philadelphia 76ers       | **1610612755** |
| Phoenix Suns             | **1610612756** |
| Portland Trail Blazers   | **1610612757** |
| Sacramento Kings         | **1610612758** |
| San Antonio Spurs        | **1610612759** |
| Toronto Raptors          | **1610612761** |
| Utah Jazz                | **1610612762** |
| Washington Wizards       | **1610612764** |


**Key Features:**
1. **Recent Performance Metrics:** Investigated the significance of a team's performance in the most recent games (`home_last5`, `vlast`, `hlast`) as predictors for upcoming playoff games.
2. **Statistical Analysis:** Conducted exploratory data analysis (EDA) to understand the distributions, correlations, and potential predictors from the dataset.
3. **Feature Engineering:** Developed features based on the win-loss records from the past games to encapsulate a team's recent momentum.

**Machine Learning Workflow:**
1. **Data Preprocessing:** Handled missing values, outliers, and transformed data into a format suitable for machine learning.
2. **Model Selection:** Chose the Random Forest Classifier due to its ability to handle large datasets with higher dimensionality and its robustness against overfitting.
3. **Model Training & Testing:** Utilized 80% of the data for training and validated the model on the remaining 20%.
4. **Results:** Achieved an accuracy score of 54% on the test data, indicating a strong potential for the model to predict playoff game outcomes based on regular season performance metrics.



nba_apiv3-master/CHANGELOG.MD
# NBA API V3 Change Log

## About Versioning

- The NBA API uses [Semantic Versioning 2.0.0](https://semver.org/)
- Dates are [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) compliant and formatted as `YYYY-MM-DD`

This package is fork `nba_api` version 1.2.1. Change log `nba_api` you can read [this](https://github.com/swar/nba_api/blob/master/CHANGELOG.MD)

# Version History

## v0.1.0

Date: 2023-09-14

- Initial commit


nba_apiv3-master/tests/unit/test_playbyplayregex.py
import pytest
from nba_apiv3.stats.library.playbyplayregex import *
from data_playbyplayregex import playbyplay


# BLOCK
@pytest.mark.parametrize('play', playbyplay['Block'])
def test_block_player(play):

    #  Validate Pattern
    assert re_block.match(play['description'])

    #  Get Capture Groups
    search_result = re_block.search(play['description'])

    #  Player Name
    player = search_result.group('player')
    assert player == play['player']

    #  Player
    block = search_result.group('blocks')
    assert block == play['blocks']


# EJECTION
@pytest.mark.parametrize('play', playbyplay['Ejection'])
def test_ejection_player(play):

    #  Validate Pattern
    assert re_ejection.match(play['description'])

    #  Get Capture Groups
    search_result = re_ejection.search(play['description'])

    #  Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Ejection Type
    ejection_type = search_result.group('ejection_type')
    assert ejection_type == play['ejection_type']


# FIELD GOAL MADE
@pytest.mark.parametrize('play', playbyplay['FieldGoalMade'])
def test_field_goal_made(play):

    # Validate Pattern
    assert re_field_goal_made.match(play['description'])

    # Get Capture Groups
    search_result = re_field_goal_made.search(play['description'])

    # Field Goal Type
    field_goal_type = search_result.group('field_goal_type')
    assert field_goal_type == play['field_goal_type']

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Disatance
    distance = search_result.group('distance')
    assert distance == play['distance']

    # Points
    points = search_result.group('points')
    assert points == play['points']

    # Player Name Assist
    player = search_result.group('player_ast')
    assert player == play['player_ast']

    # Assists
    points = search_result.group('assists')
    assert points == play['assists']


# FIELD GOAL MISSED
@pytest.mark.parametrize('play', playbyplay['FieldGoalMissed'])
def test_field_goal_missed(play):

    # Validate Pattern
    assert re_field_goal_missed.match(play['description'])

    # Get Capture Groups
    search_result = re_field_goal_missed.search(play['description'])

    # Field Goal Type
    field_goal_type = search_result.group('field_goal_type')
    assert field_goal_type == play['field_goal_type']

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Distance
    distance = search_result.group('distance')
    assert distance == play['distance']


# FOUL PLAYER
@pytest.mark.parametrize('play', playbyplay['FoulPlayer'])
def test_foul_player(play):

    # Validate Pattern
    assert re_foul_player.match(play['description'])

    # Get Capture Groups
    search_result = re_foul_player.search(play['description'])

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Foul Type
    foul_type = search_result.group('foul_type')
    assert foul_type == play['foul_type']

    # Player
    player = search_result.group('personal')
    assert player == play['personal']

    # Team Count
    team = search_result.group('team')
    assert team == play['team']

    # Referee
    referee = search_result.group('referee')
    assert referee == play['referee']


# FOUL TEAM
@pytest.mark.parametrize('play', playbyplay['FoulTeam'])
def test_foul_team(play):

    # Validate Pattern
    assert re_foul_team.match(play['description'])

    # Get Capture Groups
    search_result = re_foul_team.search(play['description'])

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Team Name
    player = search_result.group('team')
    assert player == play['team']

    # Foul Type
    foul_type = search_result.group('foul_type')
    assert foul_type == play['foul_type']

    # Referee
    referee = search_result.group('referee')
    assert referee == play['referee']


# FREE THROW MADE
@pytest.mark.parametrize('play', playbyplay['FreeThrowMade'])
def test_free_throw_made(play):

    # Validate Pattern
    assert re_free_throw_made.match(play['description'])

    # Get Capture Groups
    search_result = re_free_throw_made.search(play['description'])

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Free Throw Type
    free_throw_type = search_result.group('free_throw_type')
    assert free_throw_type == play['free_throw_type']

    # Points
    points = search_result.group('points')
    assert points == play['points']


# FREE THROW MISSED
@pytest.mark.parametrize('play', playbyplay['FreeThrowMissed'])
def test_free_throw_missed(play):

    # Validate Pattern
    assert re_free_throw_miss.match(play['description'])

    # Get Capture Groups
    search_result = re_free_throw_miss.search(play['description'])

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Free Throw Type
    free_throw_type = search_result.group('free_throw_type')
    assert free_throw_type == play['free_throw_type']


# JUMP BALL
@pytest.mark.parametrize('play', playbyplay['JumpBall'])
def test_jump_ball(play):

    # Validate Pattern
    assert re_jump_ball.match(play['description'])

    # Get Capture Groups
    search_result = re_jump_ball.search(play['description'])

    # Player Name
    player = search_result.group('player_home')
    assert player == play['player_home']

    # Player Name
    player = search_result.group('player_away')
    assert player == play['player_away']

    # Player Name
    player = search_result.group('player_tip')
    assert player == play['player_tip']


# REBOUND PLAYER
@pytest.mark.parametrize('play', playbyplay['ReboundPlayer'])
def test_rebound_player(play):

    # Validate Pattern
    assert re_rebound_player.match(play['description'])

    # Get Capture Groups
    search_result = re_rebound_player.search(play['description'])

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Offensive
    offensive = search_result.group('offensive')
    assert offensive == play['offensive']

    # Defensive
    defensive = search_result.group('defensive')
    assert defensive == play['defensive']


# REBOUND TEAM
@pytest.mark.parametrize('play', playbyplay['ReboundTeam'])
def test_rebound_team(play):

    # Validate Pattern
    assert re_rebound_team.match(play['description'])

    # Get Capture Groups
    search_result = re_rebound_team.search(play['description'])

    # Team Name
    team = search_result.group('team')
    assert team == play['team']


# STEAL
@pytest.mark.parametrize('play', playbyplay['Steal'])
def test_steal_player(play):

    # Validate Pattern
    assert re_steal.match(play['description'])

    # Get Capture Groups
    search_result = re_steal.search(play['description'])

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Steals
    offensive = search_result.group('steals')
    assert offensive == play['steals']


# SUBSTITUTION
@pytest.mark.parametrize('play', playbyplay['Substitution'])
def test_substitution(play):

    # Validate Pattern
    assert re_substitution.match(play['description'])

    # Get Capture Groups
    search_result = re_substitution.search(play['description'])

    # Player In Name
    player_in = search_result.group('player_in')
    assert player_in == play['player_in']

    # Player Out Name
    player_out = search_result.group('player_out')
    assert player_out == play['player_out']


# TIMEOUT
@pytest.mark.parametrize('play', playbyplay['Timeout'])
def test_timeout(play):

    # Validate Pattern
    assert re_timeout.match(play['description'])

    # Get Capture Groups
    search_result = re_timeout.search(play['description'])

    # Team Name
    team = search_result.group('team')
    assert team == play['team']

    # Timeout type
    timeout_type = search_result.group('timeout_type')
    assert timeout_type == play['timeout_type']

    # Full
    full = search_result.group('full')
    assert full == play['full']

    # Short
    short = search_result.group('short')
    assert short == play['short']


# TURNOVER PLAYER
@pytest.mark.parametrize('play', playbyplay['TurnoverPlayer'])
def test_turnover_player(play):

    # Validate Pattern
    assert re_turnover_player.match(play['description'])

    # Get Capture Groups
    search_result = re_turnover_player.search(play['description'])

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Turnover Type
    turnover_type = search_result.group('turnover_type')
    assert turnover_type == play['turnover_type']

    # Personal
    personal = search_result.group('personal')
    assert personal == play['personal']

    # Team
    team = search_result.group('team')
    assert team == play['team']


# TURNOVER TEAM
@pytest.mark.parametrize('play', playbyplay['TurnoverTeam'])
def test_turnover_team(play):

    # Validate Pattern
    assert re_turnover_team.match(play['description'])

    # Get Capture Groups
    search_result = re_turnover_team.search(play['description'])

    # Team Name
    team = search_result.group('team')
    assert team == play['team']

    # Turnover Type
    team = search_result.group('turnover_type')
    assert team == play['turnover_type']

    # Turnovers
    team = search_result.group('turnovers')
    assert team == play['turnovers']


# VIOLATION
@pytest.mark.parametrize('play', playbyplay['Violation'])
def test_violation(play):

    #  Validate Pattern
    assert re_violation.match(play['description'])

    #  Get Capture Groups
    search_result = re_violation.search(play['description'])

    #  Violation
    violation_type = search_result.group('violation_type')
    assert violation_type == play['violation_type']

    # Player Name
    player = search_result.group('player')
    assert player == play['player']

    # Referee
    referee = search_result.group('referee')
    assert referee == play['referee']


# VIOLATION TEAM
@pytest.mark.parametrize('play', playbyplay['ViolationTeam'])
def test_violation_team(play):

    # Validate Pattern
    assert re_violation_team.match(play['description'])

    # Get Capture Groups
    search_result = re_violation_team.search(play['description'])

    # Violation
    violation_type = search_result.group('violation_type')
    assert violation_type == play['violation_type']

    # Team Name
    team = search_result.group('team')
    assert team == play['team']


nba_apiv3-master/tests/integration/deferred_endpoints.py
import nba_apiv3.stats.endpoints as ep


class DeferredEndpoint:
    '''Simple class to represent an endpoint with deferred evaluation.'''
    def __init__(self, endpoint_class, **kwargs):
        self.endpoint_class = endpoint_class
        self.kwargs = kwargs

    def __call__(self):
        return self.endpoint_class(**self.kwargs)


# A bunch of valid but uninstantiated endpoints for testing:
deferred_endpoints = [
    DeferredEndpoint(ep.BoxScoreAdvancedV3, game_id='0021700807'),
    DeferredEndpoint(ep.BoxScoreDefensiveV2, game_id='0021700807'),
    DeferredEndpoint(ep.BoxScoreFourFactorsV3, game_id='0021700807'),
    DeferredEndpoint(ep.BoxScoreMatchupsV3, game_id='0021700807'),
    DeferredEndpoint(ep.BoxScoreMiscV3, game_id='0021700807'),
    DeferredEndpoint(ep.BoxScorePlayerTrackV3, game_id='0021700807'),
    DeferredEndpoint(ep.BoxScoreScoringV3, game_id='0021700807'),
    DeferredEndpoint(ep.BoxScoreTraditionalV3, game_id='0021700807'),
    DeferredEndpoint(ep.BoxScoreUsageV3, game_id='0021700807'),
    DeferredEndpoint(ep.PlayByPlayV3, game_id='0021700807'),
]


explore.py
import pandas as pd
import numpy as np
import os


def import_nba():
    '''
    retrieves data from .csv files.
    assigns them to variables
    '''
    nbastats = pd.read_csv('csv/nbastats_2022.csv')
    nbapo = pd.read_csv('csv/datanba_po_2022.csv')
    nbadata = pd.read_csv('csv/datanba_2022.csv')
    nbastats_po = pd.read_csv('csv/nbastats_po_2022.csv')
    pbpstats = pd.read_csv('csv/pbpstats_2022.csv')
    shotdetail = pd.read_csv('csv/shotdetail_2022.csv')
    shotdetail_po = pd.read_csv('csv/shotdetail_po_2022.csv')
    
    return nbastats, nbapo, nbadata, nbastats_po, pbpstats, shotdetail, shotdetail_po


def mvp_team():
    '''
    creates a dataframe for a time series model
    '''
    # This dictionary will be used to make a home catagory 
    # and an away catagory
    team_id_dict = {
        1610612755: "Philadelphia 76ers",
        1610612738: "Boston Celtics",
        1610612747: "Los Angeles Lakers",
        1610612744: "Golden State Warriors",
        1610612753: "Orlando Magic",
        1610612765: "Detroit Pistons",
        1610612764: "Washington Wizards",
        1610612754: "Indiana Pacers",
        1610612737: "Atlanta Hawks",
        1610612745: "Houston Rockets",
        1610612740: "New Orleans Pelicans",
        1610612751: "Brooklyn Nets",
        1610612748: "Miami Heat",
        1610612741: "Chicago Bulls",
        1610612761: "Toronto Raptors",
        1610612739: "Cleveland Cavaliers",
        1610612752: "New York Knicks",
        1610612763: "Memphis Grizzlies",
        1610612760: "Oklahoma City Thunder",
        1610612750: "Minnesota Timberwolves",
        1610612766: "Charlotte Hornets",
        1610612759: "San Antonio Spurs",
        1610612762: "Utah Jazz",
        1610612743: "Denver Nuggets",
        1610612756: "Phoenix Suns",
        1610612742: "Dallas Mavericks",
        1610612757: "Portland Trail Blazers",
        1610612758: "Sacramento Kings",
        1610612749: "Milwaukee Bucks",
        1610612746: "LA Clippers",
    }
    
    # htm dictionary to be converted into a team_id number
    htm_data = {
        "BOS": 1610612738,
        "GSW": 1610612744,
        "DET": 1610612765,
        "IND": 1610612754,
        "ATL": 1610612737,
        "BKN": 1610612751,
        "MIA": 1610612748,
        "TOR": 1610612761,
        "MEM": 1610612763,
        "MIN": 1610612750,
        "SAS": 1610612759,
        "UTA": 1610612762,
        "PHX": 1610612756,
        "SAC": 1610612758,
        "PHI": 1610612755,
        "LAL": 1610612747,
        "CHA": 1610612766,
        "WAS": 1610612764,
        "HOU": 1610612745,
        "NYK": 1610612752,
        "MIL": 1610612749,
        "DAL": 1610612742,
        "DEN": 1610612743,
        "ORL": 1610612753,
        "CLE": 1610612739,
        "CHI": 1610612741,
        "POR": 1610612757,
        "NOP": 1610612740,
        "LAC": 1610612746,
        "OKC": 1610612760
    }

    nbastats, nbapo, nbadata, nbastats_po, pbpstats, shotdetail, shotdetail_po = import_nba()
    
    # assigns the dfs i need to iterate for next step
    sdf = nbastats, nbadata, pbpstats, shotdetail
    
    # makes all columns lowercase
    for i in sdf:
        i.columns = [col.lower() for col in i.columns]
    
    # creating final dataframe from team
    team = shotdetail[['game_id', 'team_id', 'team_name', 'htm']]    
    
    # drop duplicates
    team = team.drop_duplicates(subset=['team_id','game_id'])
    
    # sorts df by game_id to be used as an index later
    team = team.sort_values('game_id')
    
    # changes the htm to a number
    team['htm'] = team['htm'].replace(htm_data)
    
    # renames to home and visitor
    team = team.rename(columns={'team_id':'visitor','htm':'home'})
    
    # drops the second copy of the game_id 
    team = team.loc[team['visitor'] != team['home']]
    
    # dropping team_name column
    team = team.drop(columns=['team_name'])
    
    # seperating final scores by game_id
    final_scores = nbadata.groupby('game_id').agg({'hs':'last', 'vs':'last'}).reset_index()
    # merging final scores onto teams
    team = team.merge(final_scores, on='game_id', how='left')
    
    # assigns 1 if home team won and 0 if home team lost
    team['result'] = np.where(team['hs'] > team['vs'], 1, 0)
    
    # set 'game_id' as the index
    team.set_index('game_id', inplace=True)
    
    # Initialize the new columns with NaN values
    team['hlast'] = np.nan
    team['vlast'] = np.nan

    # Loop through the DataFrame to populate hlast and vlast
    for idx, row in team.iterrows():
        # Get the home and visitor team ids
        home_team = row['home']
        visitor_team = row['visitor']

        # Find the last game for the home team
        prev_home_game = team[((team['home'] == home_team) | (team['visitor'] == home_team)) & (team.index < idx)]
        if not prev_home_game.empty:
            # If home team was the home team in their last game and won
            if (prev_home_game.iloc[-1]['home'] == home_team) and (prev_home_game.iloc[-1]['result'] == 1):
                team.at[idx, 'hlast'] = 1
            # If home team was the visitor team in their last game and lost
            elif (prev_home_game.iloc[-1]['visitor'] == home_team) and (prev_home_game.iloc[-1]['result'] == 0):
                team.at[idx, 'hlast'] = 1
            else:
                team.at[idx, 'hlast'] = 0

        # Find the last game for the visitor team
        prev_visitor_game = team[((team['home'] == visitor_team) | (team['visitor'] == visitor_team)) & (team.index < idx)]
        if not prev_visitor_game.empty:
            # If visitor team was the visitor team in their last game and won
            if (prev_visitor_game.iloc[-1]['visitor'] == visitor_team) and (prev_visitor_game.iloc[-1]['result'] == 0):
                team.at[idx, 'vlast'] = 1
            # If visitor team was the home team in their last game and lost
            elif (prev_visitor_game.iloc[-1]['home'] == visitor_team) and (prev_visitor_game.iloc[-1]['result'] == 1):
                team.at[idx, 'vlast'] = 1
            else:
                team.at[idx, 'vlast'] = 0

    # filling first game with losses
    # it is 16 observations
    team['hlast'].fillna(0, inplace=True)
    team['vlast'].fillna(0, inplace=True)

    # makes columns int, after i drop the nulls         
    team['hlast'] = team['hlast'].astype(int)
    team['vlast'] = team['vlast'].astype(int)

    
    # Create functions to compute rolling wins for home and visitor
    def home_last5_wins(row):
        previous_games = team[(team['home'] == row['home']) & (team.index < row.name)]
        return previous_games['result'].tail(5).sum()

    def visitor_last5_wins(row):
        previous_games = team[(team['visitor'] == row['visitor']) & (team.index < row.name)]
        return (1 - previous_games['result']).tail(5).sum()

    # Apply the functions row-wise to compute the rolling wins and assign to the desired column names
    team['home_last5'] = team.apply(home_last5_wins, axis=1)
    team['visitor_last5'] = team.apply(visitor_last5_wins, axis=1)

    # Convert to integers
    team['home_last5'] = team['home_last5'].astype(int)
    team['visitor_last5'] = team['visitor_last5'].astype(int)
    
    
    return team

    
    
    
    